lab_4/main.py

```diff
@@ -6,64 +6,236 @@
+        if not isinstance(word, str) or not word:
+            raise ValueError
+
+        try:
+            token = self.storage[word]
+        except KeyError:
+            token = len(self.storage.keys()) + 1
+            self.storage[word] = token
+        return token
 
     def get_id(self, word: str) -> int:
-        pass
+        if not isinstance(word, str) or not word:
+            raise ValueError
+
+        try:
+            token = self.storage[word]
+        except KeyError:
+            raise KeyError
+        return token
 
     def get_word(self, word_id: int) -> str:
-        pass
+        if not isinstance(word_id, int):
+            raise ValueError
+        reversed_storage = {value: key for key, value in self.storage.items()}
+        try:
+            token = reversed_storage[word_id]
+        except KeyError:
+            raise KeyError
+        return token
 
     def update(self, corpus: tuple):
-        pass
+        if not isinstance(corpus, tuple):
+            raise ValueError
+
+        for word in corpus:
+            self._put_word(word)
 
 
 def encode_text(storage: WordStorage, text: tuple) -> tuple:
-    pass
+    if not isinstance(text, tuple) or not isinstance(storage, WordStorage):
+        raise ValueError
+
+    encoded_text = []
+    for word in text:
+        encoded_text.append(storage.get_id(word))
+    return tuple(encoded_text)
 
 
 class NGramTextGenerator:
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie):
-        pass
+        self._word_storage = word_storage
+        self._n_gram_trie = n_gram_trie
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or \
+                len([w for w in context if w in self._word_storage.storage.values()]) != len(context) or\
+                len(context) < self._n_gram_trie.size - 1:
+            raise ValueError
+
+        next_word = -1
+        next_word_freq = 0
+        for n_gram, freq in self._n_gram_trie.n_gram_frequencies.items():
+            if context[:self._n_gram_trie.size-1] == n_gram[:self._n_gram_trie.size-1]:
+                if freq > next_word_freq:
+                    next_word_freq = freq
+                    next_word = n_gram[-1]
+        if next_word == -1:
+            uni = list(self._n_gram_trie.uni_grams.items())
+            uni.sort(key=lambda x: x[1], reverse=True)
+            next_word = uni[0][0][0]
+        return next_word
 
     def _generate_sentence(self, context: tuple) -> tuple:
-        pass
+        if not isinstance(context, tuple) or \
+                len([w for w in context if w in self._word_storage.storage.values()]) != len(context):
+            raise ValueError
+
+        length = 0
+        copy_context = list(context)
+        if self._word_storage.get_id("<END>") in copy_context:
+            copy_context.remove(self._word_storage.get_id("<END>"))
+        sent = copy_context
+
+        while length != 20:
```

Comment:
The number of times is pre-defined = 20, so that you always know how many times (at max) this will be executed.
This is better done with another `loop` - `for`.